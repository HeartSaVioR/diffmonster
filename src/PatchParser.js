export const LineType = {
  CONTEXT: ' ',
  ADDITION: '+',
  DELETION: '-',
};

export function parsePatch(patch) {
  const lines = patch.split('\n');
  const hunks = [];
  let hunk = null;
  for (var i = 0; i < lines.length; i++) {
    const line = lines[i];
    if (line.indexOf('@@ ') === 0) {
      if (hunk && hunk.lines.length > 0) {
        hunks.push(parseLines(hunk));
      }
      hunk = {
        position: i,
        lines: []
      };
    }
    hunk.lines.push(line);
  }
  if (hunk && hunk.lines.length > 0) {
    hunks.push(parseLines(hunk));
  }

  return hunks;
}

export function parseHunkHeader(header) {
  const result = /^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@/.exec(header);
  return {
    from: {
      start: parseInt(result[1], 10),
      count: result[2] ? parseInt(result[2], 10) : 1
    },
    to: {
      start: parseInt(result[3], 10),
      count: result[4] ? parseInt(result[4], 10) : 1
    }
  };
}

function parseLines(hunk) {
  hunk.header = hunk.lines.shift();
  hunk.range = parseHunkHeader(hunk.header);
  var fromLine = hunk.range.from.start;
  var toLine = hunk.range.to.start;
  hunk.lines = hunk.lines.map((line, i) => {
    const type = line.charAt(0);
    const position = hunk.position + i;
    const content = line.substring(1);
    switch (type) {
      case LineType.CONTEXT:
        return {
          type,
          fromLine: fromLine++,
          toLine: toLine++,
          position,
          content,
        };
      case LineType.ADDITION:
        return {
          type,
          fromLine: 0,
          toLine: toLine++,
          position,
          content,
        };
      case LineType.DELETION:
        return {
          type,
          fromLine: fromLine++,
          toLine: 0,
          position,
          content,
        };
      default:
        throw new Error('invalid format');
    }
  });
  return hunk;
}
