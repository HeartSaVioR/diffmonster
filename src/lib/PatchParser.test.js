import { parsePatch, parseHunkHeader } from './PatchParser';

const patch = "@@ -16,10 +16,14 @@\n \n package com.linecorp.armeria.common;\n \n+import static java.util.Objects.requireNonNull;\n+\n import java.util.concurrent.CompletableFuture;\n import java.util.concurrent.CompletionStage;\n import java.util.concurrent.Future;\n \n+import javax.annotation.Nullable;\n+\n /**\n  * An RPC {@link Response}. It is a {@link CompletionStage} whose result signifies the return value of an RPC\n  * call.\n@@ -29,7 +33,7 @@\n     /**\n      * Creates a new successfully complete {@link RpcResponse}.\n      */\n-    static RpcResponse of(Object value) {\n+    static RpcResponse of(@Nullable Object value) {\n         return new DefaultRpcResponse(value);\n     }\n \n@@ -41,6 +45,39 @@ static RpcResponse ofFailure(Throwable cause) {\n     }\n \n     /**\n+     * Creates a new {@link RpcResponse} that is completed successfully or exceptionally based on the\n+     * completion of the specified {@link CompletionStage}.\n+     */\n+    static RpcResponse from(CompletionStage<?> stage) {\n+        requireNonNull(stage, \"stage\");\n+        final DefaultRpcResponse res = new DefaultRpcResponse();\n+        stage.whenComplete((value, cause) -> {\n+            if (cause != null) {\n+                res.completeExceptionally(cause);\n+            } else {\n+                res.complete(value);\n+            }\n+        });\n+        return res;\n+    }\n+\n+    /**\n+     * Returns the result value if completed successfully or\n+     * throws an unchecked exception if completed exceptionally.\n+     *\n+     * @see CompletableFuture#join()\n+     */\n+    Object join();\n+\n+    /**\n+     * Returns the specified {@code valueIfAbsent} when not complete, or\n+     * returns the result value or throws an exception when complete.\n+     *\n+     * @see CompletableFuture#getNow(Object)\n+     */\n+    Object getNow(Object valueIfAbsent);\n+\n+    /**\n      * Returns the cause of the failure if this {@link RpcResponse} completed exceptionally.\n      *\n      * @return the cause, or"

describe('parsePatch', () => {
  it('parses patch', () => {
    expect(parsePatch(patch)).toMatchSnapshot();
  });
});

describe('parseHunkHeader', () => {
  it('parses header', () => {
    expect(parseHunkHeader('@@ -12,34 +56,78 @@ remainder')).toEqual({
      from: {start: 12, count: 34},
      to: {start: 56, count: 78},
    });
  });

  it('parses header with an addition and no deletions', () => {
    expect(parseHunkHeader('@@ -0,0 +1 @@ remainder')).toEqual({
      from: {start: 0, count: 0},
      to: {start: 1, count: 1},
    });
  });

  it('parses header with an addition and a deletion', () => {
    expect(parseHunkHeader('@@ -1 +1 @@ remainder')).toEqual({
      from: {start: 1, count: 1},
      to: {start: 1, count: 1},
    });
  });
});
