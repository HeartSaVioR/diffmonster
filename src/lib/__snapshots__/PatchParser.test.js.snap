exports[`parsePatch parses patch 1`] = `
Array [
  Object {
    "header": "@@ -16,10 +16,14 @@",
    "lines": Array [
      Object {
        "content": "",
        "fromLine": 16,
        "position": 1,
        "toLine": 16,
        "type": " ",
      },
      Object {
        "content": "package com.linecorp.armeria.common;",
        "fromLine": 17,
        "position": 2,
        "toLine": 17,
        "type": " ",
      },
      Object {
        "content": "",
        "fromLine": 18,
        "position": 3,
        "toLine": 18,
        "type": " ",
      },
      Object {
        "content": "import static java.util.Objects.requireNonNull;",
        "fromLine": 0,
        "position": 4,
        "toLine": 19,
        "type": "+",
      },
      Object {
        "content": "",
        "fromLine": 0,
        "position": 5,
        "toLine": 20,
        "type": "+",
      },
      Object {
        "content": "import java.util.concurrent.CompletableFuture;",
        "fromLine": 19,
        "position": 6,
        "toLine": 21,
        "type": " ",
      },
      Object {
        "content": "import java.util.concurrent.CompletionStage;",
        "fromLine": 20,
        "position": 7,
        "toLine": 22,
        "type": " ",
      },
      Object {
        "content": "import java.util.concurrent.Future;",
        "fromLine": 21,
        "position": 8,
        "toLine": 23,
        "type": " ",
      },
      Object {
        "content": "",
        "fromLine": 22,
        "position": 9,
        "toLine": 24,
        "type": " ",
      },
      Object {
        "content": "import javax.annotation.Nullable;",
        "fromLine": 0,
        "position": 10,
        "toLine": 25,
        "type": "+",
      },
      Object {
        "content": "",
        "fromLine": 0,
        "position": 11,
        "toLine": 26,
        "type": "+",
      },
      Object {
        "content": "/**",
        "fromLine": 23,
        "position": 12,
        "toLine": 27,
        "type": " ",
      },
      Object {
        "content": " * An RPC {@link Response}. It is a {@link CompletionStage} whose result signifies the return value of an RPC",
        "fromLine": 24,
        "position": 13,
        "toLine": 28,
        "type": " ",
      },
      Object {
        "content": " * call.",
        "fromLine": 25,
        "position": 14,
        "toLine": 29,
        "type": " ",
      },
    ],
    "position": 0,
    "range": Object {
      "from": Object {
        "count": 10,
        "start": 16,
      },
      "to": Object {
        "count": 14,
        "start": 16,
      },
    },
  },
  Object {
    "header": "@@ -29,7 +33,7 @@",
    "lines": Array [
      Object {
        "content": "    /**",
        "fromLine": 29,
        "position": 16,
        "toLine": 33,
        "type": " ",
      },
      Object {
        "content": "     * Creates a new successfully complete {@link RpcResponse}.",
        "fromLine": 30,
        "position": 17,
        "toLine": 34,
        "type": " ",
      },
      Object {
        "content": "     */",
        "fromLine": 31,
        "position": 18,
        "toLine": 35,
        "type": " ",
      },
      Object {
        "content": "    static RpcResponse of(Object value) {",
        "fromLine": 32,
        "position": 19,
        "toLine": 0,
        "type": "-",
      },
      Object {
        "content": "    static RpcResponse of(@Nullable Object value) {",
        "fromLine": 0,
        "position": 20,
        "toLine": 36,
        "type": "+",
      },
      Object {
        "content": "        return new DefaultRpcResponse(value);",
        "fromLine": 33,
        "position": 21,
        "toLine": 37,
        "type": " ",
      },
      Object {
        "content": "    }",
        "fromLine": 34,
        "position": 22,
        "toLine": 38,
        "type": " ",
      },
      Object {
        "content": "",
        "fromLine": 35,
        "position": 23,
        "toLine": 39,
        "type": " ",
      },
    ],
    "position": 15,
    "range": Object {
      "from": Object {
        "count": 7,
        "start": 29,
      },
      "to": Object {
        "count": 7,
        "start": 33,
      },
    },
  },
  Object {
    "header": "@@ -41,6 +45,39 @@ static RpcResponse ofFailure(Throwable cause) {",
    "lines": Array [
      Object {
        "content": "    }",
        "fromLine": 41,
        "position": 25,
        "toLine": 45,
        "type": " ",
      },
      Object {
        "content": "",
        "fromLine": 42,
        "position": 26,
        "toLine": 46,
        "type": " ",
      },
      Object {
        "content": "    /**",
        "fromLine": 43,
        "position": 27,
        "toLine": 47,
        "type": " ",
      },
      Object {
        "content": "     * Creates a new {@link RpcResponse} that is completed successfully or exceptionally based on the",
        "fromLine": 0,
        "position": 28,
        "toLine": 48,
        "type": "+",
      },
      Object {
        "content": "     * completion of the specified {@link CompletionStage}.",
        "fromLine": 0,
        "position": 29,
        "toLine": 49,
        "type": "+",
      },
      Object {
        "content": "     */",
        "fromLine": 0,
        "position": 30,
        "toLine": 50,
        "type": "+",
      },
      Object {
        "content": "    static RpcResponse from(CompletionStage<?> stage) {",
        "fromLine": 0,
        "position": 31,
        "toLine": 51,
        "type": "+",
      },
      Object {
        "content": "        requireNonNull(stage, \"stage\");",
        "fromLine": 0,
        "position": 32,
        "toLine": 52,
        "type": "+",
      },
      Object {
        "content": "        final DefaultRpcResponse res = new DefaultRpcResponse();",
        "fromLine": 0,
        "position": 33,
        "toLine": 53,
        "type": "+",
      },
      Object {
        "content": "        stage.whenComplete((value, cause) -> {",
        "fromLine": 0,
        "position": 34,
        "toLine": 54,
        "type": "+",
      },
      Object {
        "content": "            if (cause != null) {",
        "fromLine": 0,
        "position": 35,
        "toLine": 55,
        "type": "+",
      },
      Object {
        "content": "                res.completeExceptionally(cause);",
        "fromLine": 0,
        "position": 36,
        "toLine": 56,
        "type": "+",
      },
      Object {
        "content": "            } else {",
        "fromLine": 0,
        "position": 37,
        "toLine": 57,
        "type": "+",
      },
      Object {
        "content": "                res.complete(value);",
        "fromLine": 0,
        "position": 38,
        "toLine": 58,
        "type": "+",
      },
      Object {
        "content": "            }",
        "fromLine": 0,
        "position": 39,
        "toLine": 59,
        "type": "+",
      },
      Object {
        "content": "        });",
        "fromLine": 0,
        "position": 40,
        "toLine": 60,
        "type": "+",
      },
      Object {
        "content": "        return res;",
        "fromLine": 0,
        "position": 41,
        "toLine": 61,
        "type": "+",
      },
      Object {
        "content": "    }",
        "fromLine": 0,
        "position": 42,
        "toLine": 62,
        "type": "+",
      },
      Object {
        "content": "",
        "fromLine": 0,
        "position": 43,
        "toLine": 63,
        "type": "+",
      },
      Object {
        "content": "    /**",
        "fromLine": 0,
        "position": 44,
        "toLine": 64,
        "type": "+",
      },
      Object {
        "content": "     * Returns the result value if completed successfully or",
        "fromLine": 0,
        "position": 45,
        "toLine": 65,
        "type": "+",
      },
      Object {
        "content": "     * throws an unchecked exception if completed exceptionally.",
        "fromLine": 0,
        "position": 46,
        "toLine": 66,
        "type": "+",
      },
      Object {
        "content": "     *",
        "fromLine": 0,
        "position": 47,
        "toLine": 67,
        "type": "+",
      },
      Object {
        "content": "     * @see CompletableFuture#join()",
        "fromLine": 0,
        "position": 48,
        "toLine": 68,
        "type": "+",
      },
      Object {
        "content": "     */",
        "fromLine": 0,
        "position": 49,
        "toLine": 69,
        "type": "+",
      },
      Object {
        "content": "    Object join();",
        "fromLine": 0,
        "position": 50,
        "toLine": 70,
        "type": "+",
      },
      Object {
        "content": "",
        "fromLine": 0,
        "position": 51,
        "toLine": 71,
        "type": "+",
      },
      Object {
        "content": "    /**",
        "fromLine": 0,
        "position": 52,
        "toLine": 72,
        "type": "+",
      },
      Object {
        "content": "     * Returns the specified {@code valueIfAbsent} when not complete, or",
        "fromLine": 0,
        "position": 53,
        "toLine": 73,
        "type": "+",
      },
      Object {
        "content": "     * returns the result value or throws an exception when complete.",
        "fromLine": 0,
        "position": 54,
        "toLine": 74,
        "type": "+",
      },
      Object {
        "content": "     *",
        "fromLine": 0,
        "position": 55,
        "toLine": 75,
        "type": "+",
      },
      Object {
        "content": "     * @see CompletableFuture#getNow(Object)",
        "fromLine": 0,
        "position": 56,
        "toLine": 76,
        "type": "+",
      },
      Object {
        "content": "     */",
        "fromLine": 0,
        "position": 57,
        "toLine": 77,
        "type": "+",
      },
      Object {
        "content": "    Object getNow(Object valueIfAbsent);",
        "fromLine": 0,
        "position": 58,
        "toLine": 78,
        "type": "+",
      },
      Object {
        "content": "",
        "fromLine": 0,
        "position": 59,
        "toLine": 79,
        "type": "+",
      },
      Object {
        "content": "    /**",
        "fromLine": 0,
        "position": 60,
        "toLine": 80,
        "type": "+",
      },
      Object {
        "content": "     * Returns the cause of the failure if this {@link RpcResponse} completed exceptionally.",
        "fromLine": 44,
        "position": 61,
        "toLine": 81,
        "type": " ",
      },
      Object {
        "content": "     *",
        "fromLine": 45,
        "position": 62,
        "toLine": 82,
        "type": " ",
      },
      Object {
        "content": "     * @return the cause, or",
        "fromLine": 46,
        "position": 63,
        "toLine": 83,
        "type": " ",
      },
    ],
    "position": 24,
    "range": Object {
      "from": Object {
        "count": 6,
        "start": 41,
      },
      "to": Object {
        "count": 39,
        "start": 45,
      },
    },
  },
]
`;
